pydantic/_internal/_generate_schema.py

```diff
@@ -531,19 +680,47 @@ def _generate_schema(self, obj: Any) -> core_schema.CoreSchema:  # noqa: C901
         if isinstance(obj, PydanticRecursiveRef):
             return core_schema.definition_reference_schema(schema_ref=obj.type_ref)
 
-        try:
-            if obj in {bool, int, float, str, bytes, list, set, frozenset, dict}:
-                # Note: obj may fail to be hashable if it has an unhashable annotation
-                return {'type': obj.__name__}
-            elif obj is tuple:
-                return {'type': 'tuple-variable'}
-        except TypeError:  # obj not hashable; can happen due to unhashable annotations
-            pass
-
-        if obj is Any or obj is object:
-            return core_schema.any_schema()
+        return self.match_type(obj)
+
+    def match_type(self, obj: Any) -> core_schema.CoreSchema:  # noqa: C901
```

Comment:
is this method intended to be public? I know it says we'll evolve this below, but it might make sense to say something to the effect of "if you override this, know that it may be updated in future versions" or similar